certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,240 @@
+│ Invariant: length is end minus begin                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant lengthConsistency()
+    length() == absolute_diff(to_mathint(end()), to_mathint(begin()))
+    {
+        preserved {
+            requireInvariant boundariesConsistency;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: out of bounds is always 0                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant outOfBoundsDefault(int128 key)
+    (
+        key < begin() &&  // Is before beginning
+        key >= end() // Is after ending
+    ) => at_(to_uint256(key)) == 0
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: empty() is length 0 and no element exists                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant emptiness(int128 key)
+    empty() <=> length() == 0 && at_(to_uint256(key)) == 0
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: front() and back() point to the [begin, end) range limits                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant pointerCorrectness()
+    back() == at_(to_uint256(begin())) && front() == at_(to_uint256(end()) - 1)
+    {
+        preserved {
+            requireInvariant boundariesConsistency;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: pushFront adds an element at the beginning of the queue                                       │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushFront {
+    // Sanity
+    requireInvariant boundariesConsistency;
+    
+    bytes32 value;
+
+    int128 beginBefore = begin();
+    int128 beginNext = beginBefore - 1;
+    bytes32 frontBefore = front();
+
+    pushFront@withrevert(value);
+
+    int128 beginAfter = begin();
+    bytes32 frontAfter = front();
+
+    bool success = !lastReverted;
+
+    // liveness
+    assert success;
+
+    // effect
+    assert success => (
+        frontAfter == value && // Front set to value
+        beginAfter == beginNext // Begin is adjusted
```

I think this is clearer

```suggestion
        beginAfter == beginBefore - 1 // Begin is adjusted
```

---

Ended up following your advice and not checking `_begin` and `_end` pointers but proving the related invariants. This check went away during the process.